packages/angular/build/src/tools/vite/middlewares/assets-middleware.ts

```diff
@@ -76,14 +76,19 @@ export function createAngularAssetsMiddleware(
         let data: Uint8Array | string = outputFile.contents;
         if (extension === '.css') {
           // Inject component ID for view encapsulation if requested
-          const componentId = new URL(req.url, 'http://localhost').searchParams.get('ngcomp');
+          const searchParams = new URL(req.url, 'http://localhost').searchParams;
+          const componentId = searchParams.get('ngcomp');
           if (componentId !== null) {
-            // Record the component style usage for HMR updates
+            // Track if the component uses ShadowDOM encapsulation (3 = ViewEncapsulation.ShadowDom)
+            const shadow = searchParams.get('e') === '3';
+
+            // Record the component style usage for HMR updates (true = shadow; false = none; string = emulated)
             const usedIds = usedComponentStyles.get(pathname);
+            const trackingId = componentId || shadow;
             if (usedIds === undefined) {
-              usedComponentStyles.set(pathname, new Set([componentId]));
+              usedComponentStyles.set(pathname, new Set([trackingId]));
```

Comment:
It might be clearer if we structure the data differently rather than using true/false/string, as a object to make it easier to follow:

```typescript
{
    encapsulation?: string;
    componentId?: string;
}
```

I can foresee that, in the future, this can be a bit confusing whlist debugging that as component ids we store booleans

Reply:
I'm hoping we can revert this workaround sooner rather than later. Post v19 i may also look into patching the Vite client code to add support for shadow DOM with a potential followup Vite PR.